lti/views.py

```diff
@@ -0,0 +1,183 @@
+        Takes in (str)course_id (str)title (str)roles, and (obj)user.
+        Gets or creates group, and updates the user's role in that group.
+        The parameters cannot be None
+        Returns None
+        """
+        group = self.get_or_create_group(course_id=int(course_id), title=title, user=user)
+        # Determine if the user is a Teacher or Student
+        role = self.determine_role(roles)
+        # Update the roles at each launch
+        self.update_roles(user=user, group=group, role=role)
+        return None
+
+    def determine_role(self, roles):
+        """ Takes (list) roles or None, returns (str) role  """
+        if roles:
+            roles = [role.rsplit("/")[-1] for role in roles.split(",")]
+            if "Instructor" in roles:
+                return "Teacher"
+        # If roles is None, then I am returning "Student", although it is actually
+        # indicative on an LTI failure.
+        return "Student"
+
+    def get_or_create_group(self, course_id=None, title="Empty", user=None):
+        """ takes (int) course_id, and (str) title and returns a Group"""
+        try:
+            group = Group.objects.get(class_key=course_id)
+        except Group.DoesNotExist:
+            group = Group.objects.create(
+                class_key=course_id,
+                title=title,
+                created_by=user
+            )
+        return group
+
+    def update_roles(self, user, group, role):
+        """
+        Takes a user, group, and roles list.
+        Will update group's teachers or students sets
+        Returns None
+        """
+        if role == "Teacher":
+            group.teachers.add(user)
+            if user in group.students.all():
+                group.students.remove(user)
+        else:
+            group.students.add(user)
+            if user in group.teachers.all():
+                group.teachers.remove(user)
+        return None
+
+
+class LtiRegistrationView(FormView):
+
+    template_name = "lti_registration.html"
+    form_class = LtiUsernameForm
+    success_url = "/lti/lti_initializer/"
+
+    def dispatch(self, request, *args, **kwargs):
+        try:
+            request.session["lti_email"]
+        except KeyError:
+            return render(request, "lti_failure.html")
+        return super(LtiRegistrationView, self).dispatch(request, *args, **kwargs)
+
+    def form_valid(self, form):
+
+        username = form.cleaned_data["username"]
+        unique_username = form.unique_username(username)
+        if unique_username is False:
+            msg = "Sorry, that username is taken."
+            form.add_error("username", msg)
+            return super().form_invalid(form)
+        email = self.request.session["lti_email"]
+        password = get_random_alphanumeric_string(10)
+
+        new_user = get_user_model().objects.create_user(username, email=email, password=password)
+        login(self.request, new_user, backend="hedera.backends.UsernameAuthenticationBackend")
```

Comment:
Need to make sure other related objects for the user are created like `Account` and `EmailAddress`.

Reply:
From my observations, `Account` and `EmailAddress` get created during this process. There is a test to verify that an `Account` is created, but I should probably include `EmailAddress` in that test.